include/librealsense2/dds/dds-device.cpp

```diff
@@ -36,7 +74,180 @@ class dds_device::impl
+        rqos.history().kind = KEEP_LAST_HISTORY_QOS;
+        rqos.history().depth = 10;
+        rqos.endpoint().history_memory_policy
+            = eprosima::fastrtps::rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
+        _reader = DDS_API_CALL( _subscriber->create_datareader( _topic, rqos, nullptr ) );
+        if( _reader == nullptr )
+        {
+            throw std::runtime_error( "Error creating a DDS reader for " + topic_name );
+        }
+
+        LOG_DEBUG( "topic '" << topic_name << "' created" );
+    }
+
+    bool init()
+    {
+        size_t num_of_sensors = 0;
+        std::unordered_map< int , std::string> sensor_index_to_name;
+        std::unordered_map< std::string, topics::device::notification::video_stream_profiles_msg > sensor_to_video_profiles;
+        std::unordered_map< std::string, topics::device::notification::motion_stream_profiles_msg > sensor_to_motion_profiles;
+
+        // We expect to receive all of the sensors data under a timeout
+        utilities::time::timer t( std::chrono::seconds( 30) ); // TODO: refine time out
+        state_type state = state_type::WAIT_FOR_DEVICE_HEADER;
+
+        while( ! t.has_expired() && state_type::DONE != state )
+        {
+            if( _reader->wait_for_unread_message( { 1, 0 } ) )
+            {
+                dds::topics::raw::device::notification raw_data;
+                topics::device::notification data;
+                SampleInfo info;
+                // Process all the samples until no one is returned,
+                // We will distinguish info change vs new data by validating using `valid_data`
+                // field
+                while( ReturnCode_t::RETCODE_OK == _reader->take_next_sample( &raw_data, &info ) )
+                {
+                    // Only samples for which valid_data is true should be accessed
+                    // valid_data indicates that the instance is still ALIVE and the `take` return
+                    // an updated sample
+                    if( info.valid_data )
+                    {
+                        data = raw_data;
+                        auto msg_id = static_cast< topics::device::notification::msg_type >( raw_data.id() );
+                        switch( msg_id )
+                        {
+                        case topics::device::notification::msg_type::DEVICE_HEADER:
+                            if( state_type::WAIT_FOR_DEVICE_HEADER == state )
+                            {
+                                num_of_sensors = data.get< topics::device::notification::device_header_msg >()->num_of_sensors;
+                                LOG_INFO( "got DEVICE_HEADER message with " << num_of_sensors << " sensors" );
+                                state = state_type::WAIT_FOR_SENSOR_HEADER;
+                            }
+                            else
+                                LOG_ERROR( "Wrong message received, got 'DEVICE_HEADER' when the state is: " << state );
```

Comment:
Is the ordering of the messages guaranteed? I.e., is it possible to get the `SENSOR_HEADER` before `DEVICE_HEADER`?

Reply:
Yes, order is guaranteed with the chosen QoS
```
ReliabilityQosPolicyKind = RELIABLE_RELIABILITY_QOS: It indicates that the service will attempt to deliver all samples of 
the DataWriter’s history expecting an arrival confirmation from the DataReader. The data samples sent by the same 
DataWriter cannot be made available to the DataReader if there are previous samples that have not been received yet. 
The service will retransmit the lost data samples in order to reconstruct a correct snapshot of the DataWriter history 
before it is accessible by the DataReader.
```